sympy/ntheory/qs.py

```diff
@@ -0,0 +1,391 @@
+            factor_base.append(factor_base_elem(p, residue, log_p))
+    return factor_base
+
+def sieve_poly(N, M, factor_base):
+    """This step is the initialization of the 1st sieve polynomial.
+    Here `a` is selected as a product of primes such that `a` is close
+    to sqrt(2*N) / M. Other initial values of factor_base elem are
+    also intialized wich includes a_inv, B_ainv, soln1, soln2.
+
+    Parameters:
+    ===========
+
+    N : Number to be factored
+    M : sieve interval
+    factor_base : factor_base primes
+    """
+    approx_val = sqrt(2*N) / M
+    #randomly search for a combination of primes whose multiplication is close to approx_val
+    best_a, best_q, best_ratio = None, None, None
+    for _ in range(50):
+        a = 1
+        q = []
+        while(a < approx_val):
+            rand_p = 0
+            while(rand_p == 0 or rand_p in q):
+                rand_p = random.randint(3, len(factor_base) - 1)
+            p = factor_base[rand_p].prime
+            a *= p
+            q.append(rand_p)
+        ratio = a / approx_val
+        if best_ratio is None or abs(ratio - 1) < abs(best_ratio - 1):
+            best_q = q
+            best_a = a
+            best_ratio = ratio
+        if abs(best_ratio - 1) < 0.4:
+            break
+
+    a = best_a
+    q = best_q
+
+    B = []
+    for idx, val in enumerate(q):
+        q_l = factor_base[val].prime
+        gamma = factor_base[val].tmem_p * mod_inverse(a // q_l, q_l) % q_l
+        if gamma > q_l / 2:
+            gamma = q_l - gamma
+
+        B.append(a//q_l*gamma)
+
+    b = sum(B)
+    g = sieve_Polynomial([a*a, 2*a*b, b*b - N ], a, b)
+    for fb in factor_base:
+        if a % fb.prime == 0:
+            continue
+        B_ainv_i = []
+        for elem in B:
+            B_ainv_i.append(2*elem*mod_inverse(a, fb.prime) % fb.prime)
+        fb.B_ainv = B_ainv_i
+        fb.a_inv = mod_inverse(a, fb.prime)
+        fb.soln1 = (fb.a_inv*(fb.tmem_p - b)) % fb.prime
+        fb.soln2 = (fb.a_inv*(-fb.tmem_p - b)) % fb.prime
+    return g, B
+
+def ith_poly(N, factor_base, i, g, B):
+    """Initialization stage of ith poly. After we finish sieving `st polynomial
+    here we quickly change to the next polynomial from which we will again
+    start sieving.
+
+    Parameters:
+    ===========
+
+    N : number to be factored
+    factor_base : factor_base primes
+    i : integer denoting ith polynomial
+    g : (i - 1)th polynomial
+    B : array that stores a//q_l*gamma
+    """
+    from sympy import ceiling
+    v = 1
+    j = i
+    while(j % 2 == 0):
+        v += 1
+        j //= 2
+    if ceiling(i / (2**v)) % 2 == 1:
+        z = -1
+    else:
+        z = 1
+    b = g.b + 2*z*B[v - 1]
+    a = g.a
+    g = sieve_Polynomial([a*a, 2*a*b, b*b - N], a, b)
+    for fb in factor_base:
+        if a % fb.prime == 0:
+            continue
+        fb.soln1 = (fb.soln1 - z*fb.B_ainv[v - 1]) % fb.prime
+        fb.soln2 = (fb.soln2 - z*fb.B_ainv[v - 1]) % fb.prime
```

Comment:
In this function there are some variables which I think should be named properly for z, a, b, v etc. Name them so that it should be easy for us to review the code and logic.